src/main/java/com/cloudbees/jenkins/plugins/bitbucket/impl/credentials/BitbucketClientCertificateAuthenticator.java

```diff
@@ -65,27 +63,56 @@ public BitbucketClientCertificateAuthenticator(StandardCertificateCredentials cr
     }
 
     /**
-     * Sets the SSLContext for the builder to one that will connect with the selected certificate.
-     * @param context The client builder context
+     * Sets the SSLContext for the builder to one that will connect with the
+     * selected certificate.
+     *
+     * @param sslFactory The client SSL context configured in the connection
+     *        pool
      * @param host the target host name
      */
-    @Override
-    public void configureContext(HttpClientContext context, HttpHost host) {
-        try {
-            Registry<ConnectionSocketFactory> registry = RegistryBuilder.<ConnectionSocketFactory>create()
-                .register(URIScheme.HTTP.id, PlainConnectionSocketFactory.getSocketFactory())
-                .register(URIScheme.HTTPS.id, new SSLConnectionSocketFactory(buildSSLContext(), HttpsSupport.getDefaultHostnameVerifier()))
-                .build();
-            context.setAttribute(SOCKET_FACTORY_REGISTRY, registry); // override SSL registry for this context
-        } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException | KeyManagementException e) {
-            throw new BitbucketException("Failed to set up SSL context from provided client certificate", e);
-        }
+    @Restricted(NoExternalUse.class)
+    public synchronized /*required to avoid combine the same identity material twice */ void configureContext(SSLFactory sslFactory, HttpHost host) {
+        sslFactory.getKeyManager().ifPresent(baseKeyManager -> {
+            AggregatedX509ExtendedKeyManager aggregate = unwrap(baseKeyManager);
+            String routeAlias = host.getHostName();
+
+            // check if given route has been already added to the SSL context
+            Map<String, List<URI>> routes = aggregate.getIdentityRoute();
+            if (!routes.containsKey(routeAlias)) {
+                try {
+                    URI hostURI = new URI(host.toURI());
+                    routes.put(routeAlias, new ArrayList<>(List.of(hostURI)));
```

Comment:
Instead of `new ArrayList<>(List.of(hostURI))` you can do `Arrays.asList(hostURI)`